File: core/shared/src/main/scala/diplomacy/variant/standard/rule/BuildOrderGenerator.scala
package diplomacy.variant.standard.rule

import diplomacy.rule.InvalidOrderMessage
import diplomacy.board.Power

class BuildOrderGenerator[Turn_ <: Turn, Power_ <: Power] extends OrderGenerator[Turn_, Power_] {
  def ordersToSkipPhase(board: Board): Option[Set[Order]] = {
    if (board.map.powers forall { power =>
      val numOfUnits = board.units count { unit => unit.power == power }
      val numOfSupply = board.numberOfSupplyCenters.getOrElse(power, 0)

      (numOfUnits == numOfSupply) || (numOfSupply == 0)
    }) {
      Option(
        board.map.powers flatMap { power =>
          val numOfUnits = board.units count { unit => unit.power == power }
          val numOfSupply = board.numberOfSupplyCenters.getOrElse(power, 0)

          if (numOfSupply == 0) {
            board.units withFilter { unit => unit.power == power } map { unit =>
              Order.Disband(unit): Order
            }
          } else {
            Set[Order]()
          }
        }
      )
    } else {
      None
    }
  }
  def defaultOrder(board: Board)(unit: DiplomacyUnit): Order = ??? // TODO
}

File: src/variant/standard/rule/build-order-generator.ts
import { OrderGenerator } from "./order-generator"
import { StandardRuleUtils } from "./standard-rule-utils"
import { Disband } from "./order"
import { Board, Unit } from "./types"

export class BuildOrderGenerator<Power> implements OrderGenerator<Power> {
  ordersToSkipPhase (board: Board<Power>) {
    const numberOfSupplyCenters = StandardRuleUtils.numberOfSupplyCenters(board)

    const canSkip = [...board.map.powers].every(power => {
      const numOfUnits = ([...board.units].filter(unit => unit.power === power)).length
      const numOfSupply = numberOfSupplyCenters.get(power) || 0

      return (numOfUnits === numOfSupply) || (numOfSupply === 0)
    })

    if (canSkip) {
      const orders = new Set()
      board.map.powers.forEach(power => {
        const units = [...board.units].filter(unit => unit.power === power)
        const numOfSupply = numberOfSupplyCenters.get(power) || 0
        if (numOfSupply === 0) {
          units.map(unit => new Disband(unit)).forEach(o => orders.add(o))
        }
      })
      return orders
    } else {
      return null
    }
  }
  defaultOrderOf (board: Board<Power>, unit: Unit<Power>) {
    return null
  }
}
